lib/babe/epoch.go

```diff
@@ -6,137 +6,182 @@ package babe
-				randomness:     data.Randomness,
-				authorities:    data.Authorities,
-				authorityIndex: idx,
-				threshold:      b.epochData.threshold, // TODO: threshold might change if authority count changes
-			}
-		}
+	if !has {
+		logger.Criticalf("%s number=%d", errNoEpochData, epoch)
+		return nil, 0, errNoEpochData
+	}
 
-		startSlot, err = b.epochState.GetStartSlotForEpoch(epoch)
-		if err != nil {
-			return err
-		}
+	data, err := b.epochState.GetEpochData(epoch)
+	if err != nil {
+		return nil, 0, err
 	}
 
-	// if we're at genesis, we need to determine when the first slot of the network will be
-	// by checking when we will be able to produce block 1.
-	// note that this assumes there will only be one producer of block 1
-	if b.blockState.BestBlockHash() == b.blockState.GenesisHash() {
-		startSlot, err = b.getFirstSlot(epoch)
-		if err != nil {
-			return err
-		}
+	idx, err := b.getAuthorityIndex(data.Authorities)
+	if err != nil {
+		return nil, 0, err
+	}
 
-		logger.Debugf("estimated first slot as %d based on building block 1", startSlot)
-		for i := startSlot; i < startSlot+b.epochLength; i++ {
-			_, err := b.runLottery(i, epoch)
-			if err != nil {
-				if errors.Is(err, errOverPrimarySlotThreshold) {
-					continue
-				}
-				return fmt.Errorf("error running slot lottery at slot %d: error %w", i, err)
-			}
+	has, err = b.epochState.HasConfigData(epoch)
+	if err != nil {
+		return nil, 0, err
+	}
 
-			startSlot = i
-			break
+	var cfgData *types.ConfigData
+	if has {
+		cfgData, err = b.epochState.GetConfigData(epoch)
+		if err != nil {
+			return nil, 0, err
 		}
-
-		// we are at genesis, set first slot by checking at which slot we will be able to produce block 1
-		err = b.epochState.SetFirstSlot(startSlot)
+	} else {
+		cfgData, err = b.epochState.GetLatestConfigData()
 		if err != nil {
-			return err
+			return nil, 0, err
 		}
 	}
 
-	logger.Infof("initiating epoch %d with start slot %d", epoch, startSlot)
+	threshold, err := CalculateThreshold(cfgData.C1, cfgData.C2, len(data.Authorities))
+	if err != nil {
+		return nil, 0, err
+	}
 
-	for i := startSlot; i < startSlot+b.epochLength; i++ {
-		if epoch > 0 {
-			delete(b.slotToProof, i-b.epochLength) // clear data from previous epoch
-		}
+	ed := &epochData{
+		randomness:     data.Randomness,
+		authorities:    data.Authorities,
+		authorityIndex: idx,
+		threshold:      threshold,
+	}
 
-		proof, err := b.runLottery(i, epoch)
-		if err != nil {
-			if errors.Is(err, errOverPrimarySlotThreshold) {
-				continue
-			}
-			return fmt.Errorf("error running slot lottery at slot %d: error %w", i, err)
-		}
+	startSlot, err := b.epochState.GetStartSlotForEpoch(epoch)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	return ed, startSlot, nil
+}
 
-		b.slotToProof[i] = proof
-		logger.Tracef("claimed slot %d, there are now %d slots into epoch", startSlot, i-startSlot)
+func (b *Service) getLatestEpochData() (*epochData, error) {
+	var err error
+	resEpochData := &epochData{}
```

Comment:
nit this would be nicer🙏 

```suggestion
func (b *Service) getLatestEpochData() (resEpochData *epochData, err error) {
	resEpochData = &epochData{}
```